packages/compass/src/main/logging.ts

```diff
@@ -0,0 +1,227 @@
+        mongoLogId(1_001_000_003),
+        'app',
+        'Closing application'
+      );
+      writer.end();
+    });
+
+    process.on('compass:log', (meta) => {
+      writer.target.write(meta.line);
+    });
+
+    ipcMain.respondTo('compass:log', (evt, meta) => {
+      (process as EventEmitter).emit('compass:log', meta);
+    });
+
+    await manager.cleanupOldLogfiles();
+
+    return writer.logFilePath;
+  } catch (err) {
+    debug('Failure setting up logging!', err);
+  }
+}
+
+export async function extractPartialLogFile(
+  logFilePath: string,
+  tmpdir = app.getPath('temp')
+): Promise<string> {
+  const logFilename = path.basename(logFilePath, '.gz');
+  const tempFilePath = path.join(
+    tmpdir,
+    'compass_logs',
+    `compass_${logFilename}.txt`
+  );
+  debug('Extracting partial logfile', { logFilePath, tempFilePath });
+  try {
+    await fs.mkdir(path.dirname(tempFilePath), {
+      recursive: true,
+      mode: 0o700,
+    });
+    await fs.unlink(tempFilePath);
+  } catch {
+    // If there was an error from mkdir/unlink, we ignore it, if it
+    // fails for any reason other than the file not existing
+    // (which is the common case)  we'll end up running into
+    // trouble during  createWriteStream() anyway.
+  }
+
+  try {
+    await promisify(stream.pipeline)(
+      createReadStream(logFilePath),
+      // Z_SYNC_FLUSH because Compass is still running, i.e. the log
+      // file is incomplete.
+      zlib.createGunzip({ finishFlush: zlib.constants.Z_SYNC_FLUSH }),
+      createWriteStream(tempFilePath, { flags: 'wx', mode: 0o600 })
+    );
+  } catch (err) {
+    debug('Failed to extract file', err);
+    throw err;
+  }
+
+  return tempFilePath;
+}
+
+async function showLogFileDialog(logFilePath: string) {
+  const { response } = await dialog.showMessageBox({
+    type: 'info',
+    title: 'Log file for this session',
+    icon: COMPASS_ICON,
+    message: `The log file for this session can be found at ${logFilePath}`,
+    detail:
+      'Some tools may not be able to read the log file until Compass has exited.',
+    buttons: [
+      'OK',
+      'Copy to clipboard',
+      'Open Folder',
+      'Extract and open as .txt',
+    ],
+  });
+
+  switch (response) {
+    case 1:
+      clipboard.writeText(logFilePath);
+      break;
+    case 2:
+      shell.showItemInFolder(logFilePath);
+      break;
+    case 3: {
+      try {
+        const tempFilePath = await extractPartialLogFile(logFilePath);
+        shell.openItem(tempFilePath);
+      } catch (err) {
+        dialog.showErrorBox('Error extracting log file', String(err));
+      }
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+class CompassLogging {
+  private constructor() {
+    // marking constructor as private to disallow usage
+  }
+
+  private static initPromise: Promise<void> | null = null;
+
+  static async init(app: CompassApplication): Promise<void> {
+    if (this.initPromise) {
+      await this.initPromise;
```

Comment:
```suggestion
      await this.initPromise;
      return;
```

Reply:
I just came here to say the same thing - surely you mean to return here so you don't init twice?